petab/measurements.py

```diff
@@ -0,0 +1,302 @@
+    """
+    Return list of ID of parameters which occur in measurement table as
+    observable or noise parameter overrides.
+    """
+
+    def unique_preserve_order(seq):
+        seen = set()
+        seen_add = seen.add
+        return [x for x in seq if not (x in seen or seen_add(x))]
+
+    def get_unique_parameters(series):
+        return unique_preserve_order(
+            itertools.chain.from_iterable(
+                series.apply(split_parameter_replacement_list)))
+
+    return unique_preserve_order(
+        get_unique_parameters(measurement_df.observableParameters)
+        + get_unique_parameters(measurement_df.noiseParameters))
+
+
+def split_parameter_replacement_list(list_string: Union[str, numbers.Number],
+                                     delim: str = ';'
+                                     ) -> List:
+    """
+    Split values in observableParameters and noiseParameters in measurement
+    table. Convert numeric values to float.
+
+    Arguments:
+        delim: delimiter
+        list_string: delim-separated stringified list
+    """
+    if list_string is None:
+        return []
+
+    def to_float_if_float(x):
+        try:
+            return float(x)
+        except ValueError:
+            return x
+
+    if isinstance(list_string, numbers.Number):
+        # Empty cells in pandas might be turned into nan
+        # We might want to allow nan as replacement...
+        if np.isnan(list_string):
+            return []
+        return [list_string]
+
+    result = [x.strip() for x in list_string.split(delim) if len(x.strip())]
+    return [to_float_if_float(x) for x in result]
+
+
+def get_placeholders(formula_string: str, observable_id: str,
+                     override_type: str) -> set:
+    """
+    Get placeholder variables in noise or observable definition for the
+    given observable ID.
+
+    Arguments:
+        formula_string: observable formula (typically from SBML model)
+        observable_id: ID of current observable
+        override_type: 'observable' or 'noise', depending on whether `formula`
+            is for observable or for noise model
+
+    Returns:
+        (Un-ordered) set of placeholder parameter IDs
+    """
+    pattern = re.compile(
+        re.escape(override_type) + r'Parameter\d+_' + re.escape(observable_id))
+    placeholders = set()
+    for free_sym in sp.sympify(formula_string).free_symbols:
+        free_sym = str(free_sym)
+        if pattern.match(free_sym):
+            placeholders.add(free_sym)
+    return placeholders
+
+
+def create_measurement_df() -> pd.DataFrame:
+    """Create empty measurement dataframe"""
+
+    df = pd.DataFrame(data={
+        'observableId': [],
+        'preequilibrationConditionId': [],
+        'simulationConditionId': [],
+        'measurement': [],
+        'time': [],
+        'observableParameters': [],
+        'noiseParameters': [],
+        'observableTransformation': [],
+        'noiseDistribution': [],
+    })
+
```

Comment:
I would suggest either all optional fields or none (or choice of user)
`observableId | [preequilibrationConditionId] | simulationConditionId | measurement | time | [observableParameters] | [noiseParameters] | [observableTransformation] | [noiseDistribution] | [datasetId] | [replicateId]`

Reply:
done